### PR TITLE
rgw/dbstore: fixing a seg-fault when deleting an object

### PR DESCRIPTION
**Pull request by giubacc**
This fix addresses a radosgw crash when using dbstore backend and deleting an object.
Please, refer to the linked issue for details about the issue itself and how to reproduce/test it.

Fixes: https://tracker.ceph.com/issues/55225
Signed-off-by: Giuseppe Baccini <giuseppe.baccini@suse.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by dang — COMMENTED** [review]
I'm not sure this is correct, however the current code is certainly wrong.  @soumyakoduri can you take a look?

**Comment by soumyakoduri** [comment]
Thanks for reporting. This issue was fixed as part of 1c2a788 but resurfaced with latest changes to remove RGWObjCtx (88bcacc).  

As fixed earlier (https://github.com/ceph/ceph/commit/1c2a7884fa9ba53080db1213ddbd5c495cb906c5#diff-51fd0f96587f95e0f94dce76c6008b7f9a75d200b11774b8efe9fdcd082f837dL570) ,  I suggest to initialize 
*state = &(this->state);

Even in RGWRados::get_obj_state_impl(), 
>>>
 RGWObjStateManifest *sm = rctx->get_state(obj->get_obj());
  RGWObjState *s = &(sm->state);
<<<

*state refers to the one already stored as part of ObjectCtx.

File: src/rgw/store/dbstore/common/dbstore.cc [hunk 1]
```diff
@@ -1226,6 +1226,7 @@ int DB::Object::get_obj_state(const DoutPrefixProvider *dpp,
   /* XXX: For now use state->shadow_obj to store ObjectID string */
   s->shadow_obj = params.op.obj.obj_id;
 
+  *state = new RGWObjState();
```
[the remark this answers is not in the document]
Comment (soumyakoduri) [reply src/rgw/store/dbstore/common/dbstore.cc:1229]: or else similar to how Rados references to the state stored as part of ObjectCtx, maybe DBStore also can store state part of DB::Object class (something like - https://github.com/soumyakoduri/ceph/commit/249605734a0c40520034902e31fbf01a94d17e3a ..  I haven't tested the changes in detail ) & can pass its reference.

**Comment by giubacc** [comment]
Hi @dang and @soumyakoduri,
Thanks for your review; I'm currently doing some test with the hints you gave to me.
From a first sight it seems that we cannot use the same `RGWRados::get_obj_state_impl()` approach; this because `this->state` seems to be `null` at the time `**state = *s;` is being called.
I would like to try a patch that does not involve the use of the `heap` so that we can avoid the subsequent `delete(s)`.
I will keep you updated.

**Comment by giubacc** [comment]
@soumyakoduri I verified that:

```
*state = &(this->state);
```
cannot work with `dbstore` implementation.
I commited a patch that uses a temporary object `tmpstate` on the `stack` so we can avoid all the deallocating issues.

**Review by soumyakoduri — APPROVED** [review]
This works for now. I will clean up this code soon. Thanks!

File: src/rgw/store/dbstore/common/dbstore.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (giubacc) [reply src/rgw/store/dbstore/common/dbstore.cc]: This does not apply anymore with the more recent patch.

File: src/rgw/rgw_sal_dbstore.cc [hunk 2]
```diff
@@ -567,10 +567,11 @@ namespace rgw::sal {
 
   int DBObject::get_obj_state(const DoutPrefixProvider* dpp, RGWObjState **pstate, optional_yield y, bool follow_olh)
   {
-    RGWObjState* astate;
+    RGWObjState* astate = nullptr;
     DB::Object op_target(store->getDB(), get_bucket()->get_info(), get_obj());
     int ret = op_target.get_obj_state(dpp, get_bucket()->get_info(), get_obj(), follow_olh, &astate);
     if (ret < 0) {
+      delete astate;
```
[the remark this answers is not in the document]
Comment (giubacc) [reply src/rgw/rgw_sal_dbstore.cc]: This does not apply anymore with the more recent patch.

**Comment by giubacc** [comment]
superseded by https://github.com/ceph/ceph/pull/45909

**Pull request closed without merging**